File: internal/models/agent_workflow.go
package models

import "time"

type AgentWorkflowSnapshot struct {
	ActiveWorkflowID string                     `json:"active_workflow_id,omitempty"`
	Workflows        []AgentWorkflow            `json:"workflows"`
	Runs             []AgentWorkflowRun         `json:"runs"`
	SentLog          []AgentWorkflowSentLogItem `json:"sent_log,omitempty"`
	UpdatedAt        time.Time                  `json:"updated_at"`
}

type AgentWorkflow struct {
	ID          string              `json:"id"`
	Name        string              `json:"name"`
	Description string              `json:"description,omitempty"`
	Nodes       []AgentWorkflowNode `json:"nodes"`
	Edges       []AgentWorkflowEdge `json:"edges"`
	UpdatedAt   time.Time           `json:"updated_at"`
}

type AgentWorkflowNode struct {
	ID       string         `json:"id"`
	Type     string         `json:"type"`
	Label    string         `json:"label,omitempty"`
	ParentID string         `json:"parent_id,omitempty"`
	Position AgentNodePoint `json:"position"`
	Width    float64        `json:"width,omitempty"`
	Height   float64        `json:"height,omitempty"`
	Data     map[string]any `json:"data,omitempty"`
}

type AgentNodePoint struct {
	X float64 `json:"x"`
	Y float64 `json:"y"`
}

type AgentWorkflowEdge struct {
	ID           string `json:"id"`
	Source       string `json:"source"`
	SourceHandle string `json:"source_handle,omitempty"`
	Target       string `json:"target"`
	TargetHandle string `json:"target_handle,omitempty"`
	Label        string `json:"label,omitempty"`
}

type AgentWorkflowSource struct {
	ID       string  `json:"id"`
	Name     string  `json:"name"`
	Category string  `json:"category"`
	FeedURL  string  `json:"feed_url"`
	Weight   float64 `json:"weight"`
	Enabled  bool    `json:"enabled"`
}

type AgentWorkflowRun struct {
	ID             string                    `json:"id"`
	WorkflowID     string                    `json:"workflow_id,omitempty"`
	WorkflowName   string                    `json:"workflow_name,omitempty"`
	CreatedAt      time.Time                 `json:"created_at"`
	StartedAt      time.Time                 `json:"started_at,omitempty"`
	FinishedAt     time.Time                 `json:"finished_at,omitempty"`
	Status         string                    `json:"status,omitempty"`
	Summary        string                    `json:"summary"`
	OutputText     string                    `json:"output_text,omitempty"`
	Items          []AgentWorkflowItem       `json:"items"`
	NodeResults    []AgentWorkflowNodeResult `json:"node_results,omitempty"`
	FetchErrors    []AgentRunError           `json:"fetch_errors,omitempty"`
	DeliveryErrors []AgentRunError           `json:"delivery_errors,omitempty"`
}

type AgentWorkflowNodeResult struct {
	NodeID   string         `json:"node_id"`
	NodeType string         `json:"node_type"`
	Status   string         `json:"status"`
	Summary  string         `json:"summary,omitempty"`
	Metadata map[string]any `json:"metadata,omitempty"`
}

type AgentWorkflowItem struct {
	Title       string `json:"title"`
	URL         string `json:"url"`
	SourceID    string `json:"source_id"`
	SourceName  string `json:"source_name"`
	PublishedAt string `json:"published_at,omitempty"`
	Summary     string `json:"summary,omitempty"`
}

type AgentWorkflowSentLogItem struct {
	URLNormalized string    `json:"url_normalized"`
	SentAt        time.Time `json:"sent_at"`
	Title         string    `json:"title"`
}

type AgentRunError struct {
	Target string `json:"target"`
	Error  string `json:"error"`
}

File: web/admin/src/lib/agent-workflow.ts
import { request } from './api';

export type AgentWorkflowSource = {
  id: string;
  name: string;
  category: string;
  feed_url: string;
  weight: number;
  enabled: boolean;
};

export type AgentWorkflowNode = {
  id: string;
  type: string;
  label?: string;
  parent_id?: string;
  position: { x: number; y: number };
  width?: number;
  height?: number;
  data?: Record<string, unknown>;
};

export type AgentWorkflowEdge = {
  id: string;
  source: string;
  source_handle?: string;
  target: string;
  target_handle?: string;
  label?: string;
};

export type AgentWorkflowDefinition = {
  id: string;
  name: string;
  description?: string;
  nodes: AgentWorkflowNode[];
  edges: AgentWorkflowEdge[];
  updated_at: string;
};

export type AgentWorkflowNodeResult = {
  node_id: string;
  node_type: string;
  status: string;
  summary?: string;
  metadata?: Record<string, unknown>;
};

export type AgentWorkflowRun = {
  id: string;
  workflow_id?: string;
  workflow_name?: string;
  created_at: string;
  started_at?: string;
  finished_at?: string;
  status?: string;
  summary: string;
  output_text?: string;
  items: Array<Record<string, unknown>>;
  node_results?: AgentWorkflowNodeResult[];
  fetch_errors?: Array<{ target: string; error: string }>;
  delivery_errors?: Array<{ target: string; error: string }>;
};

export type AgentWorkflowSnapshot = {
  active_workflow_id: string;
  workflows: AgentWorkflowDefinition[];
  runs: AgentWorkflowRun[];
  sent_log?: Array<Record<string, unknown>>;
  updated_at: string;
};

export function saveAgentWorkflow(payload: AgentWorkflowSnapshot) {
  return request('/agent/workflow', { method: 'PUT', body: JSON.stringify(payload) });
}

export function runAgentWorkflow(workflowId?: string) {
  return request<AgentWorkflowRun>('/agent/workflow/run', {
    method: 'POST',
    body: JSON.stringify({ workflow_id: workflowId ?? '' }),
  });
}

export function normalizeAgentWorkflowSnapshot(input: AgentWorkflowSnapshot | null | undefined): AgentWorkflowSnapshot {
  const workflow = input ?? ({} as AgentWorkflowSnapshot);
  return {
    ...workflow,
    workflows: arrayOrEmpty(workflow.workflows),
    runs: arrayOrEmpty(workflow.runs),
    sent_log: arrayOrEmpty(workflow.sent_log),
  };
}

function arrayOrEmpty<T>(value: T[] | null | undefined): T[] {
  return Array.isArray(value) ? value : [];
}
